html/modules/custom/reliefweb_subscriptions/src/AmazonSesMessageBuilder.php

```diff
@@ -0,0 +1,81 @@
+<?php
+
+namespace Drupal\reliefweb_subscriptions;
+
+use Drupal\amazon_ses\MessageBuilder;
+
+/**
+ * Extend the Amazon SES module's message builder.
+ */
+class AmazonSesMessageBuilder extends MessageBuilder {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildMessage(array $message) {
+    // Adjust headers and expected message properties to work with the base
+    // amazon_ses message builder.
+    if (isset($message['headers'])) {
+      foreach ($message['headers'] as $key => $value) {
+        switch (strtolower($key)) {
+          case 'from':
+            $message['from'] = $value;
+            unset($message['headers'][$key]);
+            $message['headers']['From'] = $value;
+            break;
+
+          case 'reply-to':
+            $message['reply-to'] = $value;
+            unset($message['headers'][$key]);
+            $message['headers']['Reply-to'] = $value;
+            break;
+
+          case 'cc':
+            unset($message['headers'][$key]);
+            $message['headers']['Cc'] = $value;
+            break;
+
+          case 'bcc':
+            unset($message['headers'][$key]);
+            $message['headers']['Bcc'] = $value;
+            break;
+        }
+      }
+    }
+
+    /** @var Symfony\Component\Mime\Email $email */
+    $email = parent::buildMessage($message);
+
+    // Add extra headers.
+    // @see \Drupal\reliefweb_subscriptions\ReliefWebSubscriptionsMailer::generateEmail().
+    if (isset($message['params']['headers'])) {
+      /** @var Symfony\Component\Mime\Headers $headers */
+      $headers = $email->getHeaders();
+      foreach ($message['params']['headers'] as $header => $value) {
+        $headers->addHeader($header, $value);
+      }
+    }
```

Comment:
I don't understand this bit; the headers are added to the $headers class but that's not used again after that?

Reply:
`$headers` is a reference to the Headers object of the `$email` object. So we are adding extra headers directly to the email's headers.

Those headers are then used to form the raw email when the `$email` is converted to string: https://git.drupalcode.org/project/amazon_ses/-/blob/3.0.x/src/AmazonSesHandler.php?ref_type=heads#L95